src/spack_commands.py

```diff
@@ -26,35 +26,38 @@ def log_with_spack(command: str,
     log_file = Path(
         spack_c2sm_path) / 'log' / machine_name() / test_category / filename
 
+    if allow_srun and getpass.getuser() == 'jenkins':
+        # Wrap the command with srun.
+        # The '-c' argument should be in sync with
+        # sysconfig/<machine>/config.yaml config:build_jobs for max efficiency
+        sruns = {
+            'daint': 'srun -t 02:00:00 -C gpu -A g110 -c 12 -n 1',
+            'tsa': 'srun -t 02:00:00 -c 6',
+        }
+        if machine_name() in sruns:
+            srun = sruns[machine_name()]
+            if uenv:
+                srun += f' --uenv={uenv}:/user-environment'
+
+            command = f'{srun} {command}'
+
     # Setup spack env
     spack_env = f'. {spack_c2sm_path}/setup-env.sh'
     if uenv:
         spack_env += ' /user-environment'
 
-    # Only use srun as Jenkins user and not on balfrin
-    use_srun = allow_srun and getpass.getuser() == 'jenkins' and machine_name(
-    ) != 'balfrin'
-
-    uenv_args = ''
-    if uenv:
-        uenv_mount_point = f'{uenv}:/user-environment'
-        if use_srun:
-            uenv_args = '--uenv=' + uenv_mount_point
-        else:
-            uenv_args = 'squashfs-mount ' + uenv_mount_point + ' -- '
+    if env:
+        spack_env += f'; spack env activate -d {env}'
 
-    # Distribute work with 'srun'
-    if use_srun:
-        # The '-c' argument should be in sync with
-        # sysconfig/<machine>/config.yaml config:build_jobs for max efficiency
+    # Redirect output to log file.
+    # This is done on the command line level to avoid buffering and potentially losing buffered content.
+    # Without parentheses, srun errors would not be captured.
+    # '2>&1' redirects stderr to stdout.
+    command = f'{spack_env}; ({command}) >> {log_file} 2>&1'
 
-        # No entry for balfrin, trigger and error instead if requested to run with srun
-        srun = {
-            'daint': 'srun -t 02:00:00 -C gpu -A g110 -c 12 -n 1',
-            'tsa': 'srun -t 02:00:00 -c 6',
-        }[machine_name()]
-    else:
-        srun = ''
+    # Mount user environment
+    if uenv:
+        command = f'squashfs-mount {uenv}:/user-environment -- bash -c "{command}"'
```

Comment:
Don't you mount the uenv now twice?
Once as an argument to srun command in L. 40 and then here?

`squashfs-mount` and `--uenv` should not be used simultanously I guess.